### DISCUSSION TITLE
Can we send ResponseFinal before Query dropping?

### DISCUSSION THREAD

**Discussion by pojiro** [eclipse-zenoh/roadmap#117] [Q&A] [answered]
Currenty I'm developing Zenohex, Zenoh Elixir binding.
https://github.com/b5g-ex/zenohex/tree/pojiro/0.2.0-dev

We use Zenoh Query data on ErlangVM which Elixir runs on. Elixir doesn't have block scope, so we cannot control query's drop timing. Erlang VM's GC can release them, but we can't control the timing. ( In fact, we can by calling `:erlang.garbage_collect()`, but it's too much for  just sending ResponseFinal. 

Therefore we can't control the timing of send_response_final which is called when query drops.
https://github.com/eclipse-zenoh/zenoh/blob/0.10.1-rc/zenoh/src/queryable.rs#L55-L63

So I would like to ask, is there way to use send_response_final before query dropping?

**Comment by p-avital** [comment, marked as answer]
Hi there,

To my knowledge, there isn't (nor should be) a way to do that. In general, Zenoh relies on destructors a lot, so I wouldn't recommend trying to bypass destructors, but rather see how you can ensure that they are called properly by your wrapper.

The way we've done that in the Kotlin and Python bindings is that we generally store provide the binding objects with `Option<T>` where `T` is the type that has useful drop behaviour. Then, the implementation of `close()` for that binding object calls a rust function that calls `Option::take`, allowing `T` to be dropped properly.

I'm not very familiar with Elixir, but if you want to avoid forcing the user to remember calling `query.send_final()`, a few GC'd language have ways to register calls to happen when an object's strong count reaches 0: for example, JS has `FinalizationRegistry`.

**Reply by pojiro** [reply]
@p-avital Thank you for your quick answer.

Based your info, I could implement a function, like `query.send_final()`, which drops query using `Option::take`.
Thanks for explaing me how. I'm new to Rust and couldn't figure out how to do that. Thank you!!
